### DISCUSSION TITLE
Allow control over playback

### DISCUSSION THREAD

**Discussion by wpower12** [Quefumas/gensound#14] [Ideas]
Hi! Just started using gensound as part of a demo project. I was trying to use it as part of a 'procedurally generated sound effects' system, and was wondering if you could expose more control in the play method. Right now it just loops, I think? and I was hoping to have an interface similar to PyGames mixer, where you can pass a value to play of -1 to loop, or some integer for 'number of plays'. 

Right now to get it to work, im trying to pull the actual audio numpy array out of the resulting Signal object, and pass that into mixer, but that's leading to some weird issues I can't figure out. Being able to control all the playback just through gensound would be amazing.

**Comment by Quefumas** [comment]
Hello and welcome!

This is actually two questions, in a sense, so I'll try to answer them separately, then suggest some solutions.

The first, relating to the playback method: currently Gensound relies on [SimpleAudio](https://github.com/hamiltron/py-simple-audio) for playback, which was chosen for being lightweight and cross-platform. I don't imagine Gensound will ever try to handle the hardware aspects of the playback by itself, but the plan is to make it as easy as possible for users to delegate the playback and I/O into whatever other library they may have available, so in theory the user could easily choose to direct the audio to any of simpleaudio, PyGame, etc., and also supply any library-specific arguments on the way (allowing the audio to loop, if the library supports it). However, that may still take a while to implement, and currently it doesn't seem like there's a convenient way to loop directly from simpleaudio (currently the audio plays exactly once, and may be interrupted).

Second, I agree that there's a need for an easier interface to get the actual resulting audio so it can be passed elsewhere in the code. In fact, I'm working on that right now, and it will be part of version 0.4 that I hope to release within a week.

Although this feature would help, it still wouldn't make it as convenient as you'd hope, but I can offer these options which should be already functional right now:
* If the Signal before playback is `s`, you can repeat it as many times as you like using the syntax `s**100` (to repeat a 100 times), then call `play()`. Note that this means that the Signal will be computed anew a 100 times, which could be slow (come to think of it this is kinda silly and can be changed quite easily).
* The raw audio generated from the Signal is accessible in two formats: by calling `Signal.realise(sample_rate)` you get an Audio object, whose `audio` property is a `numpy.ndarray` object of shape `(num_channels, num_samples)` and data type is `numpy.float64`. I imagine that this data by itself would not be playable by most libraries, but if you go on to call the Audio object's `_prepare_buffer(self, byte_width, max_amplitude)` , this will attach to the Audio object a property named `buffer`, which is essentially a bytearray which should be directly playable, when accompanied with the sample rate and byte width used (I recommend using a byte width of 2 for the moment). The buffer array holds integers of the appropriate width and endianness rather than floats, and also has the samples for channels interleaved.
* If all else fails, perhaps you can output the audio as a WAV file, and use the OS (manually or through `os` module) to play the file in a loop.

Example code for the 2nd suggestion:
```python
s = Sine("A4", 1e3)
audio = s.realise(sample_rate=44100)
bytes = audio._prepare_buffer(byte_width=2, max_amplitude=0.9)
```
The variables `bytes` should be passable without much fuss to most other playback interfaces. In 0.4 there will be an easier interface for this process exactly. Also be careful since the behavior of the `max_amplitude` argument will change slightly in 0.4.

Please let me know if this helps!

**Reply by wpower12** [reply]
Thank you so much for the detailed response! I'm VERY new to audio and synth in general, so this was fantastic. I just changed my script to the following  I was able to pass the buffer into the pygame mixer methods w.o issue and get the right tone to play. 
```python
audio = Sine('A4', duration=1e3).realise(sample_rate=44100)
audio._prepare_buffer(byte_width=2, max_amplitude=0.9)
samples = audio.buffer
pg.mixer.Sound.__init__(self, buffer=samples)
```
Thanks again! Look forward to the updates!

**Reply by Quefumas** [reply]
Great! I'm glad that worked. When 0.4 comes out make sure to update that bit.

(Also yes, my bad, `_prepare_buffer` doesn't actually return anything, your code is the correct one.)

**Comment by wpower12** [comment]
In case you're interested, here's a quick example of using gensound as part of a pygame.mixer.Sound object. Might be fodder for an examples section of your wiki?

```python
# An example class to use with the pygame sound system, the mixer.
# When an object of this class is instantiated, it will be able
# to be '.play()'ed, as other example Sounds would be. 
import pygame as pg

from gensound            import Sine, Sawtooth
from gensound.curve      import SineCurve
from gensound.transforms import ADSR, Amplitude

class ExampleGSTone(pg.mixer.Sound):
	def __init__(self):
		m_freq = pg.mixer.get_init()[0]
		
		# Initial Chord
		signal =  0.4*Sawtooth('C4', duration=1e3)
		signal += 0.3*Sawtooth('E4', duration=1e3)
		signal += 0.3*Sawtooth('G4', duration=1e3)

		# Some low Freq Modulation
		signal *= Amplitude(SineCurve(frequency=2, depth=0.2, baseline=0.7, duration=1e3))
		
		# An ADSR Envelope. 
		signal *= ADSR(attack=0.002e3, decay=0.03e3, sustain=0.8, release=0.1e3)

		# 'Realise' the signal into an actual audio sample object.
		audio = signal.realise(sample_rate=m_freq) 
		audio._prepare_buffer(byte_width=2, max_amplitude=0.9)

		# We can now use the prepared buffer as the input to the
		# pygame mixer Sound class.
		pg.mixer.Sound.__init__(self, buffer=audio.buffer)

```

**Reply by Quefumas** [reply]
Nice! It's really cool to see what people come up with. I'm also glad it's clear enough how to use everything (though some of the features here aren't even been documented yet, I think???)

BTW, there is/will be a dedicated Transform for amplitude modulation, `SineAM(frequency, size)`, which is simpler to use, though it's not documented yet and needs some work.

Also remember you can use both channels to make stuff interesting, for example:
```python
from gensound import Pan, SineAM

signal =  0.4*Sawtooth('C4', duration=1e3)
signal += 0.3*Sawtooth('E4', duration=1e3)
signal += 0.3*Sawtooth('G4', duration=1e3)

stereo_signal = signal*Pan(-100) + signal*Pan(100)
stereo_signal[0] *= Amplitude(SineCurve(frequency=2, depth=0.2, baseline=0.7, duration=1e3))
stereo_signal[1] *= Amplitude(SineCurve(frequency=1.37, depth=0.2, baseline=0.7, duration=1e3))

stereo_signal *= ADSR(attack=0.002e3, decay=0.03e3, sustain=0.8, release=0.1e3)
```

**Comment by Quefumas** [comment]
Hi @wpower12 ,

In case it's still relevant for you (or for anybody else reading), the new version 0.5 supports playback directly to PyGame. You don't have to do anything - just call `Signal.play()` as usual, and Gensound will automatically detect PyGame is installed and take care of the rest.

Any keyword arguments you add will be carried over to PyGame as well, so you can use the `loop` argument as usual. Also, the `play()` function will return a `pygame.mixer.Sound` object, which you can later `stop()` (or do other things with).

Have fun!
